File: local-media-downloader/backend/app/utils/timecode.py
"""Parse and validate HH:MM:SS / MM:SS timecodes for clip ranges."""
from __future__ import annotations

import re

_PATTERN = re.compile(r"^(?:(\d{1,2}):)?([0-5]?\d):([0-5]\d)$")


def parse_timecode(value: str) -> float:
    """Return the timecode as seconds. Raises ValueError if malformed."""
    match = _PATTERN.match(value.strip())
    if not match:
        raise ValueError(
            f"Invalid timecode '{value}'. Expected HH:MM:SS or MM:SS."
        )
    hours_str, minutes_str, seconds_str = match.groups()
    hours = int(hours_str) if hours_str else 0
    minutes = int(minutes_str)
    seconds = int(seconds_str)
    return hours * 3600 + minutes * 60 + seconds


def validate_clip_range(start: str, end: str) -> tuple[float, float]:
    start_s = parse_timecode(start)
    end_s = parse_timecode(end)
    if end_s <= start_s:
        raise ValueError("Clip end time must be after start time.")
    return start_s, end_s

File: local-media-downloader/frontend/src/utils/timecode.ts
const PATTERN = /^(?:(\d{1,2}):)?([0-5]?\d):([0-5]\d)$/;

export function parseTimecode(value: string): number {
  const match = PATTERN.exec(value.trim());
  if (!match) {
    throw new Error(`Invalid timecode '${value}'. Expected HH:MM:SS or MM:SS.`);
  }
  const hours = match[1] ? parseInt(match[1], 10) : 0;
  const minutes = parseInt(match[2], 10);
  const seconds = parseInt(match[3], 10);
  return hours * 3600 + minutes * 60 + seconds;
}
